File: Final/Code/Project_4/DFS/General_Graph/General_Graph.py
from collections import defaultdict

class Graph:
    def __init__(self):
        self.adj = defaultdict(list)

    def add_edge(self, u, v):
        self.adj[u].append(v)
        # Nếu là đồ thị vô hướng, thêm cạnh ngược:
        # self.adj[v].append(u)

    def dfs_util(self, v, visited):
        visited.add(v)
        print(v, end=' ')
        for neighbor in self.adj[v]:
            if neighbor not in visited:
                self.dfs_util(neighbor, visited)

    def dfs(self):
        visited = set()
        for vertex in self.adj:
            if vertex not in visited:
                self.dfs_util(vertex, visited)

g = Graph()
g.add_edge(0, 1)
g.add_edge(0, 2)
g.add_edge(1, 2)
g.add_edge(2, 0)
g.add_edge(2, 3)
g.add_edge(3, 3)

print("DFS của đồ thị:")
g.dfs()

File: Final/Code/Project_4/DFS/General_Graph/General_Graph.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>

using namespace std;

class Graph {
private:
    unordered_map<int, vector<int>> adj;

    void dfs_util(int v, unordered_set<int>& visited) {
        visited.insert(v);
        cout << v << " ";
        for (int neighbor : adj[v]) {
            if (visited.find(neighbor) == visited.end()) {
                dfs_util(neighbor, visited);
            }
        }
    }

public:
    void add_edge(int u, int v) {
        adj[u].push_back(v);
        // adj[v].push_back(u);
    }

    void dfs() {
        unordered_set<int> visited;
        for (auto& pair : adj) {
            int vertex = pair.first;
            if (visited.find(vertex) == visited.end()) {
                dfs_util(vertex, visited);
            }
        }
    }
};

int main() {
    Graph g;
    g.add_edge(0, 1);
    g.add_edge(0, 2);
    g.add_edge(1, 2);
    g.add_edge(2, 0);
    g.add_edge(2, 3);
    g.add_edge(3, 3);

    cout << "DFS của đồ thị: ";
    g.dfs();
    cout << endl;

    return 0;
}